scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

```diff
@@ -0,0 +1,527 @@
+package com.twitter.scalding.typed.memory_backend
+
+import cascading.flow.{ FlowDef, FlowConnector }
+import cascading.pipe.Pipe
+import cascading.tap.Tap
+import cascading.tuple.{ Fields, TupleEntryIterator }
+import com.twitter.scalding.typed._
+import com.twitter.scalding.{ Config, Execution, ExecutionCounters, Mode }
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ ArrayList, Collections, UUID }
+import scala.concurrent.{ Future, ExecutionContext => ConcurrentExecutionContext, Promise }
+import scala.collection.mutable.{ ArrayBuffer, Map => MMap }
+import scala.collection.JavaConverters._
+
+import Execution.{ ToWrite, Writer }
+
+class AtomicBox[T <: AnyRef](init: T) {
+  private[this] val ref = new AtomicReference[T](init)
+
+  def lazySet(t: T): Unit =
+    ref.lazySet(t)
+
+  /**
+   * use a pure function to update the state.
+   * fn may be called more than once
+   */
+  def update[R](fn: T => (T, R)): R = {
+
+    @annotation.tailrec
+    def loop(): R = {
+      val init = ref.get
+      val (next, res) = fn(init)
```

Comment:
If I understood correctly the purpose of this class, caching doesn't seem possible since `fn` depends on the current value and needs to be computed on each retry (the value will be different on each retry). @johnynek maybe renaming `loop` to `retry` would be more clear?